tests/components/mqtt/test_init.py

```diff
@@ -64,619 +58,642 @@ def mock_mqtt():
+    await hass.async_block_till_done()
+    assert calls_a.called
 
-        unsub = mqtt.subscribe(self.hass, "test/state", None)
-        mqtt.subscribe(self.hass, "test/state", None)
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"]._mqttc.subscribe.called
+    calls_b = MagicMock()
+    await mqtt.async_subscribe(hass, "test/state", calls_b)
+    await hass.async_block_till_done()
+    assert calls_b.called
 
-        unsub()
-        self.hass.block_till_done()
-        assert not self.hass.data["mqtt"]._mqttc.unsubscribe.called
 
-    def test_restore_subscriptions_on_reconnect(self):
-        """Test subscriptions are restored on reconnect."""
-        # Fake that the client is connected
-        self.hass.data["mqtt"].connected = True
+async def test_not_calling_unsubscribe_with_active_subscribers(
+    hass, mqtt_client_mock, mqtt_mock
+):
+    """Test not calling unsubscribe() when other subscribers are active."""
+    # Fake that the client is connected
+    mqtt_mock.connected = True
 
-        mqtt.subscribe(self.hass, "test/state", None)
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"]._mqttc.subscribe.call_count == 1
+    unsub = await mqtt.async_subscribe(hass, "test/state", None)
+    await mqtt.async_subscribe(hass, "test/state", None)
+    await hass.async_block_till_done()
+    assert mqtt_client_mock.subscribe.called
 
-        self.hass.data["mqtt"]._mqtt_on_disconnect(None, None, 0)
-        self.hass.data["mqtt"]._mqtt_on_connect(None, None, None, 0)
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"]._mqttc.subscribe.call_count == 2
+    unsub()
+    await hass.async_block_till_done()
+    assert not mqtt_client_mock.unsubscribe.called
 
-    def test_restore_all_active_subscriptions_on_reconnect(self):
-        """Test active subscriptions are restored correctly on reconnect."""
-        # Fake that the client is connected
-        self.hass.data["mqtt"].connected = True
-
-        self.hass.data["mqtt"]._mqttc.subscribe.side_effect = (
-            (0, 1),
-            (0, 2),
-            (0, 3),
-            (0, 4),
-        )
 
-        unsub = mqtt.subscribe(self.hass, "test/state", None, qos=2)
-        mqtt.subscribe(self.hass, "test/state", None)
-        mqtt.subscribe(self.hass, "test/state", None, qos=1)
-        self.hass.block_till_done()
+async def test_restore_subscriptions_on_reconnect(hass, mqtt_client_mock, mqtt_mock):
+    """Test subscriptions are restored on reconnect."""
+    # Fake that the client is connected
+    mqtt_mock.connected = True
 
-        expected = [
-            call("test/state", 2),
-            call("test/state", 0),
-            call("test/state", 1),
-        ]
-        assert self.hass.data["mqtt"]._mqttc.subscribe.mock_calls == expected
+    await mqtt.async_subscribe(hass, "test/state", None)
+    await hass.async_block_till_done()
+    assert mqtt_client_mock.subscribe.call_count == 1
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5737505)